containerd-shim-spin/src/engine.rs

```diff
@@ -182,57 +183,74 @@ impl SpinEngine {
-                let redis_trigger: RedisTrigger = self
-                    .build_spin_trigger(working_dir, app, app_source)
-                    .await
-                    .context("failed to build spin trigger")?;
-
-                info!(" >>> running spin trigger");
-                redis_trigger.run(spin_trigger::cli::NoArgs)
-            }
-            SqsTrigger::TRIGGER_TYPE => {
-                let sqs_trigger: SqsTrigger = self
-                    .build_spin_trigger(working_dir, app, app_source)
-                    .await
-                    .context("failed to build spin trigger")?;
-
-                info!(" >>> running spin trigger");
-                sqs_trigger.run(spin_trigger::cli::NoArgs)
-            }
-            _ => {
-                todo!("Only Http, Redis and SQS triggers are currently supported.")
-            }
-        };
+        let mut futures_list = Vec::with_capacity(trigger_types.len());
+        for trigger_type in trigger_types.iter() {
+            let f = match trigger_type.to_owned() {
+                HttpTrigger::TRIGGER_TYPE => {
+                    let http_trigger: HttpTrigger = self
+                        .build_spin_trigger(working_dir.clone(), app.clone(), app_source.clone())
+                        .await
+                        .context("failed to build spin trigger")?;
+
+                    info!(" >>> running spin http trigger");
+                    http_trigger.run(spin_trigger_http::CliArgs {
+                        address: parse_addr(SPIN_ADDR).unwrap(),
+                        tls_cert: None,
+                        tls_key: None,
+                    })
+                }
+                RedisTrigger::TRIGGER_TYPE => {
+                    let redis_trigger: RedisTrigger = self
+                        .build_spin_trigger(working_dir.clone(), app.clone(), app_source.clone())
+                        .await
+                        .context("failed to build spin trigger")?;
+
+                    info!(" >>> running spin redis trigger");
+                    redis_trigger.run(spin_trigger::cli::NoArgs)
+                }
+                SqsTrigger::TRIGGER_TYPE => {
+                    let sqs_trigger: SqsTrigger = self
+                        .build_spin_trigger(working_dir.clone(), app.clone(), app_source.clone())
+                        .await
+                        .context("failed to build spin trigger")?;
+
+                    info!(" >>> running spin trigger");
+                    sqs_trigger.run(spin_trigger::cli::NoArgs)
+                }
+                _ => {
+                    todo!("Only Http, Redis and SQS triggers are currently supported.")
+                }
+            };
+
+            futures_list.push(f)
+        }
+
         info!(" >>> notifying main thread we are about to start");
-        f.await
+
+        // exit as soon as any of the trigger completes/exits
+        let (result, _, rest) = future::select_all(futures_list).await;
```

Comment:
Why as soon as one completes?  Would that mean if the result of one of the others failed we woudln't know about it?  should we be waiting for all triggers finish, or some type of queue to handle the various results?

Reply:
Hello James, 

this is to keep it in sync with the [Spin behavior](https://github.com/fermyon/spin/blob/86a58771091c02eaf40b5395a2a6bc711740f038/src/commands/up.rs#L222)

Reply:
could we add a log to print which trigger failed / terminates?

Reply:
👍 

```
time="2024-04-04T02:58:27.951588261Z" level=info msg=" >>> trigger type 'redis' exited"
```

Reply:
@jsturtevant - does that answer your question? (re: exit behavior). thanks.

Reply:
> this is to keep it in sync with the [Spin behavior](https://github.com/fermyon/spin/blob/86a58771091c02eaf40b5395a2a6bc711740f038/src/commands/up.rs#L222)

Interesting!  I guess we should do the same as spin cli, thanks!